c10/util/Metaprogramming.h

```diff
@@ -17,12 +18,24 @@ namespace c10 { namespace guts {
 template<class Func> struct function_traits {
   static_assert(!std::is_same<Func, Func>::value, "In function_traits<Func>, Func must be a plain function type.");
 };
-template<class Result, class... Args>
-struct function_traits<Result (Args...)> {
-  using func_type = Result (Args...);
+template<class Result>
+struct function_traits<Result ()> {
+  using func_type = Result ();
   using return_type = Result;
-  using parameter_types = typelist::typelist<Args...>;
-  static constexpr auto number_of_parameters = sizeof...(Args);
+  using parameter_types = typelist::typelist<>;
+  static constexpr auto number_of_parameters = 0;
+  using first_arg_is_dispatchKeySet = std::is_same<DispatchKeySet, Result>;
```

Comment:
this doesn't belong into `c10/util` since it's PyTorch specific and `c10/util` are general utilities independent from pytorch internals. Instead of this, you can check it at the call site using `std::is_same<DispatchKeySet, typelist::head_t<func_traits::parameter_types>>`

Also note that you seem to be comparing `Result` here while you claim to compare the first argument type, that's likely not what you want. More correct would be to just set `first_arg_is_dispatchKeySet` to `std::false_type`. But this likely doesn't matter if you follow the paragraph above and remove the changes here.

Reply:
Definitely agreed.

One small issue I ran into doing this refactor: When I blindly use `std::is_same<DispatchKeySet, typelist::head_t<func_traits::parameter_types>>` and `!std::is_same<DispatchKeySet, typelist::head_t<func_traits::parameter_types>>` as my method of specialization, functions with 0 arguments break/complain because `typelist::head_t` throws a static assert.

My fix (not pushed yet) was to add a new `typelist::head_with_default_t` that returns whatever default type you pass in if the list is empty, since my goal is really just to return false instead of failing. I might be able to alternatively write a specialization for functors with an empty argument list, but I couldn't come up with a way to do that cleanly.

Reply:
`head_with_default_t` sounds good to me